torch/csrc/jit/argument_spec.h

```diff
@@ -153,10 +153,16 @@ struct ArgumentSpec {
 
  private:
   TypePtr fillType(TypePtr original, size_t& offset) const {
-    if (original->isSubtypeOf(TensorType::get())) {
+    if (original->isSubtypeOf(TensorType::get()) ||
+        original->isSubtypeOf(OptionalType::ofTensor())) {
       auto& arg = args.at(offset++);
-      if (!arg.defined())
-        return AutogradZeroTensorType::get();
+      if (!arg.defined()) {
+        if (original->isSubtypeOf(OptionalType::ofTensor())) {
```

Comment:
Could you add a comment here?

Reply:
Ok, this actually looks wrong to me. Basically when we return an `Optional[Tensor]` from forward, then there are 3 cases to consider for gradients:
1. The result was `None`. It's not a tensor, so it cannot be differentiated.
2. The result was unused/gets no gradient, so it should be treated as a zero tensor.
3. The result was used/gets a gradient, so it is non-zero and we shouldn't optimize it out.

This `if` does not seem to distinguish 1. and 2., which later results in some very awkward code you had to put into AD code, and I would very much like to see those parts gone. A good way to start fixing it is to start differentiating between them in here.

Reply:
This is now fixed to distinguish between all three. :)